File: solution/463IslandPerimeter/463IslandPerimeter.cpp
#include <vector>
using namespace std;

class Solution {
public:
  int islandPerimeter(vector<vector<int>> &grid) {
    int n = grid.size();
    if (n == 0)
      return 0;
    int m = grid[0].size();
    if (m == 0)
      return 0;
    static const int offset_x[] = {-1, 1, 0, 0};
    static const int offset_y[] = {0, 0, -1, 1};
    int sum = 0;
    for (int j = 0; j < n; ++j) {
      for (int i = 0; i < m; ++i) {
        if (grid[j][i] == 1) {
          for (int k = 0; k < 4; ++k) {
            int y = j + offset_y[k], x = i + offset_x[k];
            if (x >= 0 && x < m && y >= 0 && y < n) {
              if (grid[y][x] == 0) {
                ++sum;
              }
            } else {
              ++sum;
            }
          }
        }
      }
    }
    return sum;
  }
};

int main() { return 0; }

File: src/463IslandPerimeter/463IslandPerimeter.rs
impl Solution {
    pub fn island_perimeter(grid: Vec<Vec<i32>>) -> i32 {
        let m = grid.len();
        if m == 0 {return 0;}
        let n = grid[0].len();
        if n == 0 {return 0;}
        let mut result = 0;
        for j in 0..m {
            for i in 0..n {
                if grid[j][i]!=0 {
                    let top  = if j>0 && grid[j-1][i]==1 {1} else {0};
                    let left = if i>0 && grid[j][i-1]==1 {1} else {0};
                    let right= if i<n-1 && grid[j][i+1]==1 {1} else {0};
                    let bottom=if j<m-1 && grid[j+1][i]==1 {1} else {0};
                    result += 4-top-left-right-bottom;
                }     
            }
        }
        result
    }
}
